### ISSUE TITLE
BareMetalHost deprovisioning phase hooks 

### ISSUE THREAD

**Issue by fmuyassarov** [metal3-io/baremetal-operator#1105]
In some cases, deletion of CAPI Machine doesn't have time to drain the node, because as soon as deletionTimestamp is added on the BareMetalHost, BMO/Ironic takes the host through deprovisioning. As such, Machine is hanging in drain and the BareMetalHost is already deprovisioned. We would like to have a way to delay the deprovisioning of the host until Machine draining has not been completed. One way we could achieve is via hooks.

This is motivated from Cluster API [Machine Deletion Hooks](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200602-machine-deletion-phase-hooks.md), which allows some custom components to take some custom actions on the Machine after it has been marked for the deletion but before being drained and/or associated VM/host is terminated.

The form of the hook could be either finalizer or annotation (preferable) which will pause the deprovisioning of the host, until we ensure that the draining is completed. Once completed, we can remove the finalizer or annotation to let the BareMetalHost continue with deprovisioning.

/cc @zaneb @dtantsur @kashifest @andfasano 

/kind feature

**Comment by knfoo** [comment]
I am not sure if this is relevant or not.
But  we are running fairly large BMH so it be cool to have a way to mark the machine for deletion that will just cordon the node and wait till the workload have remove it self from the node because of usual deployment patterns.
Added to that have a way to say that the node should be drained after x-time if workload have not moved it self off the BMH

**Comment by zaneb** [comment]
I am confused about the scenario in which this problem is triggered.

CAPM3 should _never_ be deleting a BareMetalHost.

**Comment by fmuyassarov** [comment]
I'm sorry for not being clear. But here is the scenario and I should have properly described it in the description.

Let’s assume in case when there is CAPI + CAPM3 involved, and Machine/Metal3Machine/BMH is provisioned and the user deletes the BMH. We normally don’t recommend deleting BMH when it is owned by CAPI/CAPM3 CRs. Instead, the user should delete the Machine (top to bottom) which eventually brings the BMH into the Available state. But if this happens, then manual intervention is needed to bring the cluster back to a proper healthy state. Because when BMH deletion is requested without deleting the Machine, then deprovisioning starts right away, without drain or etcd checks that would be normally done by CAPI. To avoid that, deprovisioning of BMH should be blocked if there was no request to delete the owning Machine (i.e. if there is no `deletionTimestamp` on the Machine), even though the user tried to delete the BMH.

Now assume that the user tries to delete the Machine normally, BMH deprovisioning kicks off. Once finished, Machine and Metal3Machine will be deleted and BMH is in Available state. At this point, BMO could delete the BMH because it has a `deletionTimestamp` set.

In a nutshell, we want to block the deletion of BMH that has `ownerReference` (Metal3Machine) and we want to block starting of deprovisioning if there was no request to the Machine deletion even though there was a request to delete the BMH. BMH can be deleted only after it reaches Available state or in other words when it doesn’t have `ownerReference` in its metadata.

**Comment by fmuyassarov** [comment]
@zaneb @knfoo please let me know about your thoughts.

**Comment by zaneb** [comment]
> @zaneb I just checked the code in CAPM3, and CAPM3 [sets ownerReference](https://github.com/metal3-io/cluster-api-provider-metal3/blob/d219aabc19aa91115dfb77b1b7a4cd59b6e43cb3/baremetal/metal3machine_manager.go#L356) on the BMH when it is being provisioned. When deletion of Metal3Machine is requested, CAPM3 controller [removes the consumerRef](https://github.com/metal3-io/cluster-api-provider-metal3/blob/d219aabc19aa91115dfb77b1b7a4cd59b6e43cb3/baremetal/metal3machine_manager.go#L688) and removes [ownerReference](https://github.com/metal3-io/cluster-api-provider-metal3/blob/d219aabc19aa91115dfb77b1b7a4cd59b6e43cb3/baremetal/metal3machine_manager.go#L691).

That makes so sense, it should just never set the owner reference.

**Comment by zaneb** [comment]
I worked on some similar issues in our (ancient) fork here: https://github.com/openshift/cluster-api-provider-baremetal/pull/113

If the Host is deprovisioned while it is allocated, we [transistion the Machine to the Failed phase](https://github.com/openshift/cluster-api-provider-baremetal/commit/3217bcbfb687720b5e0ade8be6346097ce72f2dc). Previously we used to delete the Machine, but it looks like we concluded with the Cluster API experts that the right thing to do if a Machine is failed is leave it there for debugging if no MHC is set, or let MHC clean it up if the user has enabled it.

The node is drained and then deleted when either the user or MHC deletes the Machine. I did fix a downstream issue where MHC was not correctly deleting the Machine but instead trying to reboot it: https://github.com/openshift/machine-api-operator/pull/688 - I'm not sure whether that bug also exists in the Cluster API or not.


We could say that we're not going to deprovision while the ConsumerRef is set, but that's a pretty big change in behaviour (e.g. would definitely break CAPBM, which doesn't unset its ConsumerRef when it sees ahost being deleted), and it would be very difficult to communicate to the user why the Host is just sitting there in Provisioned when they've already tried to delete it (in fact, if you delete it from kubectl, it will hang forever).

I think the answer should be "if you don't want to have to manually intervene to recover after you've manually intervened to do some crazy thing, then you should enable MachineHealthCheck".

**Comment by kashifest** [comment]
> > @zaneb I just checked the code in CAPM3, and CAPM3 [sets ownerReference](https://github.com/metal3-io/cluster-api-provider-metal3/blob/d219aabc19aa91115dfb77b1b7a4cd59b6e43cb3/baremetal/metal3machine_manager.go#L356) on the BMH when it is being provisioned. When deletion of Metal3Machine is requested, CAPM3 controller [removes the consumerRef](https://github.com/metal3-io/cluster-api-provider-metal3/blob/d219aabc19aa91115dfb77b1b7a4cd59b6e43cb3/baremetal/metal3machine_manager.go#L688) and removes [ownerReference](https://github.com/metal3-io/cluster-api-provider-metal3/blob/d219aabc19aa91115dfb77b1b7a4cd59b6e43cb3/baremetal/metal3machine_manager.go#L691).
> 
> That makes so sense, it should just never set the owner reference.

I believe the reason we used to set owner reference is to ensure the owner reference chain for clusterctl for [pivoting](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).

**Comment by zaneb** [comment]
Oh right, the pivot.

**Comment by zaneb** [comment]
> Our case is a bit unique because leaving the Machine in Failed state, when MHC enabled, will not save us anyways. MHC will try to remediate the unhealthy host and after the maximum number of retries it will fall back to the deletion.

I think either the MHC or the remediation controller needs to take that into account. If the host has been deprovisioned then rebooting will never work and shouldn't be tried.

As I mentioned above, I fixed this in OpenShift with https://github.com/openshift/machine-api-operator/pull/688. In that version of the Machine API, if the Machine is Failed then the controller never calls the actuator again (and it is the actuator that does the external remediation - in our case reboot), so the only correct thing to do is delete the Machine. In the latest CAPI there might be an argument that this decision should be left to the remediation controller, but in any event something must decide that the Machine is unrecoverable and delete it. @slintes or @n1r1 might have more context on that.

**Comment by kashifest** [comment]
> > Our case is a bit unique because leaving the Machine in Failed state, when MHC enabled, will not save us anyways. MHC will try to remediate the unhealthy host and after the maximum number of retries it will fall back to the deletion.
> 
> I think either the MHC or the remediation controller needs to take that into account. If the host has been deprovisioned then rebooting will never work and shouldn't be tried.
> 
> As I mentioned above, I fixed this in OpenShift with [openshift/machine-api-operator#688](https://github.com/openshift/machine-api-operator/pull/688). In that version of the Machine API, if the Machine is Failed then the controller never calls the actuator again (and it is the actuator that does the external remediation - in our case reboot), so the only correct thing to do is delete the Machine. In the latest CAPI there might be an argument that this decision should be left to the remediation controller, but in any event something must decide that the Machine is unrecoverable and delete it. @slintes or @n1r1 might have more context on that.

MHC is a recovery process. What @fmuyassarov is proposing , is a first hand prevention mechanism. 

Do you see any issue @zaneb if we add the annotation in BMH from CAPM3 and leaving all the BMO workflow as it is and only react to prevent the BMH from unwanted/accidental de-provisioning once CAPM3 has set the annotation?

**Comment by zaneb** [comment]
> Do you see any issue @zaneb if we add the annotation in BMH from CAPM3 and leaving all the BMO workflow as it is and only react to prevent the BMH from unwanted/accidental de-provisioning once CAPM3 has set the annotation?

I mean, yeah, I think it's a horrible user experience.

And I don't think we should protect the users from this. There's no way to undo a delete in k8s, so if the user deletes something we should let it get removed instead of sitting around in an undead state with the DeletionTimestamp set. And other stuff that references it should observe that and report their status appropriately (in this case the Machine should be Failed, just as would happen if you deleted a VM instance out from under a Machine provisioned with a cloud's CAPI provider).

**Comment by kashifest** [comment]
> And I don't think we should protect the users from this. There's no way to undo a delete in k8s, so if the user deletes something we should let it get removed instead of sitting around in an undead state with the DeletionTimestamp set. And other stuff that references it should observe that and report their status appropriately (in this case the Machine should be Failed, just as would happen if you deleted a VM instance out from under a Machine provisioned with a cloud's CAPI provider).

I don't think the idea is understood correctly by one of us. I understood that the idea is not trying to undo a delete in k8s, the idea is allowing to delay BMH deprovisioning until Machine drains gracefully. 

I agree that in absence of this logic normally Machine should be in a failed state to correctly depict the scenario, which we already have currently. The problem is we don't have any mechanism to recover Machine from that state and this situation is unique to metal3 which is why the proposal is trying to solve the issue in metal3 only.

According to me this idea is pretty close to `[Finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/)`  and I dont see it as a horrible user experience since this is a common concept. @fmuyassarov have you explored if Finalizers can be a way to solve this issue if annotation is something that is not popular in the community ? (Although I still think annotation should solve this easily)

**Comment by zaneb** [comment]
> this situation is unique to metal3

As discussed in the meeting, I don't think it is. Deleting a VM from a cloud provider (using the cloud APIs directly) is exactly equivalent to deprovisioning a BMH. The Metal3Machine ought to detect this and cause the Machine to be marked as Failed, as other CAPI providers do afaik.

**Comment by zaneb** [comment]
Yeah, that bug could explain a lot of the problems you've been having.

However, that appears to be due to a limitation in CAPM3. It's a mistake to assume that the Machine is failed only if the BMH is removed. It's actually failed as soon as the BMH starts deprovisioning (for any reason). In OpenStack terms, you've lost your machine not when the Ironic Node is deleted but when the Nova Server that has provisioned it has been deleted. [Decomposing the BMH API](https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit#heading=h.shaa5llsd7ov) would make this more obvious. Until then, the "instance" that is comparable to other CAPI providers is a _combination_ of the BMH and the image it is provisioned with.

Originally CAPBM did not detect deprovisioning and I assume that flowed through into the CAPM3 implementation. I fixed it in the OpenShift fork of CAPBM here: https://github.com/openshift/cluster-api-provider-baremetal/commit/3217bcbfb687720b5e0ade8be6346097ce72f2dc. So how it should actually work is more like:

1. user deletes the node
2. host enters deprovisioning state (ironic node deprovisioning kicks off)
5. capi machine enters failed state
6. if MHC is enabled, it will delete the failed machine. If not, user deletes the machine
7. metal3machine is also deleted
8. a new metal3machine is created followed by the machine creation by MD/KCP
3. host enters deleting state (ironic node deletion kicks off)
4. both ironic node and host are deleted

**Comment by fmuyassarov** [comment]
> Yeah, that bug could explain a lot of the problems you've been having.
> 
> However, that appears to be due to a limitation in CAPM3. It's a mistake to assume that the Machine is failed only if the BMH is removed. It's actually failed as soon as the BMH starts deprovisioning (for any reason).
> 
this makes sense

**Comment by fmuyassarov** [comment]
Hi @zaneb @kashifest 
Actually even after the fix with BMO deletion, we still have another issue with Machine deletion. Deletion could be blocked and I've tested it. If there is some workload on the host (a daemonset with some volume attached) Machine can not be deleted, because as said earlier CAPI tries to drain the node, which doesn't exist because BMH got deleted. We could just pause this deletion for a few seconds so that CAPM3 actually does something on the Machine so that draining succeeds. Compared to previous discussions, we are not silently blocking the deletion of BMH but rather taking some actions on Machine objects to lower the risk of getting stuck with unhealthy machines. The action that CAPM3 takes could be: 

1. bmh is created and a finalizer/annotation is added by CAPM3
2. user deletes the host, deletionTimestamp is added
3. de-provisoning is blocked due to existence of finalizer/annotation on the host
4. CAPM3 noticed deletionTimestamp on bmh but not on the machine, so it will issue machine deletion
5. once a machine has deletionTimestamp, CAPM3 removes the finalizer/annotation from the bmh, 
6. de-provisioning is unblocked

alternative
1. bmh is created and a finalizer/annotation is added by CAPM3
2. user deletes the host, deletionTimestamp is added
3. de-provisoning is blocked due to existence of finalizer/annotation on the host
4. CAPM3 noticed deletionTimestamp on bmh but not on the machine, so it will set the Machine into Failed state
5. user deletes the machine manually, deletionTimestamp is added on the machine
6. once a machine has deletionTimestamp, CAPM3 removes the finalizer/annotation from the bmh
7. de-provisioning is unblocked

As said, we are not blocking deletion of BMH here, but doing pre-deletion steps to ensure we don't run into a non-recoverable state of Machine.
